src/ui/marker.js

```diff
@@ -437,22 +435,17 @@ export default class Marker extends Evented {
             return;
         }
         const mapLocation = map.unproject(pos);
-        let opacity = 1;
-        if (map._usingGlobe()) {
-            opacity = this._occluded(mapLocation) ? 0 : 1;
-        } else if (map.transform._terrainEnabled() && map.getTerrain()) {
-            opacity = this._occluded(mapLocation) ? TERRAIN_OCCLUDED_OPACITY : 1;
+        let opacity = (map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat)) ? 0 : 1;
+        if (opacity) {
+            if (map.transform._terrainEnabled() && map.getTerrain()) {
+                opacity = this._behindTerrain() ? TERRAIN_OCCLUDED_OPACITY : 1;
+            }
+            opacity *= (1.0 - map._queryFogOpacity(mapLocation));
```

Comment:
nit: Let's try to be more sequential and reduce overly branching.

```suggestion
        const behindGlobe = map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat);
        const behindTerrain = map.transform._terrainEnabled() && map.getTerrain() && this._behindTerrain();

        const fogOcclusion = 1.0 - map._queryFogOpacity(mapLocation);

        const opacity = (behindGlobe ? 0 : (behindTerrain ? TERRAIN_OCCLUDED_OPACITY : 1)) * fogOcclusion;
```

Reply:
One concern with your rewrite is that checks for terrain and opacity will run even when the marker is occluded by the globe, shouldn't we skip those if possible?

It's possible to keep that optimization while rewriting these lines as a series of nested ternary operators, but IMO readability is worsened by doing this:

```js
        const behindGlobe = map._usingGlobe() && isLngLatBehindGlobe(map.transform, this._lngLat);
        const opacity = behindGlobe ? 0 :
            (map.transform._terrainEnabled() && map.getTerrain() && this._behindTerrain() ? TERRAIN_OCCLUDED_OPACITY : 1) *
            (1.0 - map._queryFogOpacity(mapLocation));
```